src/database-postgresql.cpp

```diff
@@ -0,0 +1,505 @@
+	case PGRES_COMMAND_OK:
+		break;
+	case PGRES_TUPLES_OK:
+		if (PQntuples(m_res))
+			s = std::string(PQgetvalue(m_res, 0,0), PQgetlength(m_res, 0, 0));
+		break;
+	default:
+		errorstream << "PostgreSQL Error:" << PQresultErrorMessage(m_res) << std::endl;
+		break;
+	}
+
+	PQclear(m_res);
+	return s;
+}
+
+const char * Database_PostgreSQL::posTochar(const v3s16 &pos)
+{
+	char buffer[33];
+	snprintf(buffer, 32, "%ld", getBlockAsInteger(pos));
+
+	const char* ret = buffer;
+	return ret;
+}
+
+void Database_PostgreSQL::createDatabase()
+{
+	DSTACK(__FUNCTION_NAME);
+
+	verifyDatabase();
+
+	m_res = PQexec(m_conn,
+		"CREATE TABLE IF NOT EXISTS blocks ("
+			"pos BIGINT PRIMARY KEY,"
+			"data BYTEA"
+		");"
+	);
+
+	switch(PQresultStatus(m_res)) {
+	case PGRES_COMMAND_OK:
+		infostream << "ServerMap: Postgres database structure was created" << std::endl;
+		break;
+	default:
+		errorstream << "PostgreSQL Error:" << PQresultErrorMessage(m_res) << std::endl;
+	       	throw BaseException(PQresultErrorMessage(m_res));
+		break;
+	}
+
+	PQclear(m_res);
+	return;
+}
+
+bool Database_PostgreSQL::ping()
+{
+	if (!m_conninfo.length())
+		connectString();
+
+	switch(PQping(m_conninfo.c_str())) {
+	case PQPING_OK:
+		return true;
+		break;
+	case PQPING_REJECT:
+		errorstream << "PostgreSQL Error: The server is running but is in a state that "
+				<< "disallows connections (startup, shutdown, or crash recovery)."
+				<< std::endl;
+		break;
+	case PQPING_NO_RESPONSE:
+		errorstream << "PostgreSQL Error: The server could not be contacted. This might "
+				<< "indicate that the server is not running, or that there is "
+				<< "something wrong with the given connection parameters (for "
+				<< "example, wrong port number), or that there is a network connectivity "
+				<< "problem (for example, a firewall blocking the connection request)."
+				<< std::endl;
+		break;
+	case PQPING_NO_ATTEMPT:
+		errorstream << "PostgreSQL Error: No attempt was made to contact the server, because "
+				"the supplied parameters were obviously incorrect or there was some "
+				"client-side problem (for example, out of memory)." << std::endl;
+		break;
+
+	}
+
+	return false;
+}
+
+void Database_PostgreSQL::listAllLoadableBlocks(std::vector<v3s16> &dst)
+{
+	DSTACK(__FUNCTION_NAME);
+
+	int numrows;
+	int row;
+
+	m_res = PQexecPrepared(m_conn, "read_pos", 0, NULL, NULL, NULL, 1); 
+
+	switch(PQresultStatus(m_res)) {
+	case PGRES_COMMAND_OK:
+		break; // No data returned
+	case PGRES_TUPLES_OK:
+
+		numrows = PQntuples(m_res);
+
+		if (numrows < 1)
+			break;
+
+		for (row = 0; row < numrows; row++) {
+			v3s16 p = getIntegerAsBlock(stoi64(PQgetvalue(m_res, row, 0)));
+			dst.push_back(p);
+		}
+
+		break;
+	default:
+		errorstream << "PostgreSQL Error:" << PQresultErrorMessage(m_res) << std::endl;
+	       	throw BaseException(PQresultErrorMessage(m_res));
+		break;
+	}
+
+	PQclear(m_res);
+	return;
+}
+
+Database_PostgreSQL::~Database_PostgreSQL()
+{
+	if (m_res)
```

Comment:
What do these checks try to accomplish? m_res is never set to NULL anywhere, even if PQclear has been invoked on it before. same for m_conn.